Add getProperyNames() functions to Block and Indexed Block.

ricrogz opened the pull request:
Addresses #41 

Using these functions, we can recover a list of the properties stored in the block.

Given the way properties are stored, duplicated names are allowed if they have different types, so, to disambiguate, it is required to provide a hint of the data type each property is associated. These hints are provided as the initial of the data type: 'b' for boolean, 'i' for integer, 'r' for real number, and 's' for string.

Also, a couple of unrelated improvements: making Block m_name const, use std::move to initialize Block and IndexedBlock m_names, and delete the copy constructor and assignment operator.

lorton commented:
I think we access internal structures to get around this in the Writer, perhaps we should switch to using this there?

ricrogz commented:
No, I think it is ok to directly access the data members inside the writer; since we have access to them, it is easy to iterate on them.

This change was meant for external code, which doesn't have access, and so, any property that is not hardcoded at compile time or provided by the user could not be accessed.

In fact, I am thinking it makes more sense to split my function and retrieve property names for each data type separately, since it will make it easier to handle them.

lorton commented:
Ah yeah that makes sense to let the writer keep doing what it's doing.

ricrogz commented:
As I said, I reconsidered: for an external user, it will probably be easier to handle 4 different functions to get the names of each property, since we provide different functions to handle each type of data.

This also leaves it to the user to deal with duplicate property names.

lorton commented:
Hi @ricrogz - but that still doesn't affect the internal use does it?  Or should we switch just to use an API instead of internals as a good principle?

ricrogz commented:
No, still no change required: the new functions (as the one I initially proposed) store the properties in vectors, while the auxiliaries we use for writing dump them out straight to the output stream. Using the new functions would only add an unrequired intermediary step to what we are doing now.

ricrogz commented:
I moved the refactoring to a different PR, and implemented Dan's suggestion of using a templated method to forward const references to the property maps.

ricrogz commented:
> I like this better. Are you cool with it?

Totally!
